README.md
[![npm version](https://badge.fury.io/js/sd-builder.svg)](https://badge.fury.io/js/sd-builder)
[![Dependency Status](https://david-dm.org/staticdeploy/sd-builder.svg)](https://david-dm.org/staticdeploy/sd-builder)
[![devDependency Status](https://david-dm.org/staticdeploy/sd-builder/dev-status.svg)](https://david-dm.org/staticdeploy/sd-builder#info=devDependencies)

# sd-builder

Opinionated builder for react web projects.

## Usage

* `sd-builder dev` sets up dev environment with auto-recompiling
* `sd-builder build` builds the project
* `sd-builder config` writes the app configuration to `app-config.js`

## Main conventions

### Html

* The builder expects there to be an `app/main.html` file, which is compiled
  into `build/index.html`. For now the compilation consists in a simple a
  copy/paste.

### JS

* The builder expects there to be an `app/main.jsx` file, which is used as
  `webpack`'s entry point. The generated bundle is written to
  `build/_assets/js/app.js`.

* Files are compiled by babel. It's up to the user to specify which plugins to
  use by installing and listing them in a `.babelrc` config file.

* npm modules listed in `deps.json`'s `js` array are separated from the main
  bundle and compiled into `build/_assets/js/vendor.js`.

* During `webpack`'s compilation `NODE_PATH` includes the `app` directory.

* It's possible to `require` / `import` `.js`, `.jsx` and `.json` files.

* Source maps are always generated.

### Assets

* The content of the folder `app/assets` will be recursively copied into
  `build/_assets`.

### CSS

* Files listed in `deps.json`'s `css` array are bundled (concat) into
  `build/_assets/css/vendor.css`.

### Fonts

* Files listed in `deps.json`'s `fonts` array are copied (concat) into
  `build/_assets/fonts`.

### Version

* the build generates a `VERSION.txt` file (written to `build/VERSION.txt`)
  with the following format:
  * if building in a git repository, `[package.json version] - [git commit sha]`
  * otherwise, `[package.json version]`

### Changelog

* when a `CHANGELOG.md` file is present, it's copied to `build/CHANGELOG.md`

## Configuration

Running `sd-builder config` a `build/app-config.js` file is generated, exporting
one global variable, `window.APP_CONFIG`, which is a map of key-value pairs
gathered from:

* the `.env` file when `NODE_ENV=development`
* environment variables prefixed by `__APP_CONFIG__` file when
  `NODE_ENV=production`

You should add `app-config.js` script in your `main.html` file.

```html
<script src="app-config.js"></script>
```

## Build customization via build-time environment variables

At build time, the following environment variables can be used to customize the
build:

* `NODE_ENV`: defaults to `development`
* `EXEC_ENV`: defaults to `browser`

The variables are:

* passed to [gulp-preprocess](https://github.com/jas/gulp-preprocess) when
  building `main.html`
* passed to webpack `DefinePlugin` when building js files

Example `main.html` build customization:

```html
<!-- @if EXEC_ENV=='cordova' -->
<!--
    This ends up in the compiled index.html only when, in the build
    environment, process.env.EXEC_ENV === "cordova"
-->
<script src="cordova.js"></script>
<!-- @endif -->
```

Example js build customization:

```js
console.log(process.env.EXEC_ENV);
/*
*   When in the build environment process.env.EXEC_ENV === "cordova", the above
*   line of code will be compiled into the line `console.log("cordova");`
*/
```

When `NODE_ENV=production` JS and CSS files are minified.


CHANGELOG.md
## 4.0.1 (2017-04-23)

Fix:

* fix config task if build folder not exists

## 4.0.0 (2016-10-15)

Features:

* build `CHANGELOG.md` file

Breaking changes:

* renamed version file `VERSION` to `VERSION.txt`

## 3.0.1 (2016-09-13)

Fixes:

* add missing dependency `yargs`

## 3.0.0 (2016-09-13)

Breaking changes:

* no more `lint`, `test`, `coverage` commands
* only runs with nodejs >= 6


src/tasks/dev.js
const browserSync = require("browser-sync");
const history = require("connect-history-api-fallback");
const gulp = require("gulp");

const {APP_DIR, BUILD_DIR, DEPS_PATH} = require("../config");
const build = require("./build");
const config = require("./config");

function setupDevServer () {
    browserSync({
        server: {
            baseDir: BUILD_DIR,
            middleware: [history()]
        },
        files: `${BUILD_DIR}/**/*`,
        port: 8080,
        ghostMode: false,
        injectChanges: false,
        notify: false,
        open: false,
        reloadDebounce: 1000
    });
}

function setupWatchers () {
    gulp.watch(`${APP_DIR}/main.html`, build.mainHtml);
    gulp.watch(`${APP_DIR}/.env`, config);
    gulp.watch([`${APP_DIR}/**/*.jsx`, `${APP_DIR}/**/*.js`], build.allScripts);
    gulp.watch(`${APP_DIR}/assets/**/*`, build.appAssets);
    gulp.watch(DEPS_PATH, () => Promise.all([
        build.allScripts(true),
        build.vendorFonts(),
        build.vendorStyles()
    ]));
}

module.exports = function dev () {
    build();
    config();
    setupDevServer();
    setupWatchers();
};


src/tasks/build.js
const {promisify} = require("bluebird");
const {execSync} = require("child_process");
const fs = require("fs-extra");
const gulp = require("gulp");
const gulpLoadPlugins = require("gulp-load-plugins");
const proGulp = require("pro-gulp");
const webpack = require("webpack");

const gp = gulpLoadPlugins();

/*
*   Constants
*/

const {
    NODE_ENV,
    EXEC_ENV,
    MINIFY_FILES,
    APP_DIR,
    BUILD_DIR,
    DEPS_PATH
} = require("../config");

/*
*   Utils
*/

function getCommitSha () {
    try {
        return execSync("git rev-parse HEAD").toString();
    } catch (ignore) {
        console.warn("Failed to get commit sha via git command");
    }
    try {
        return execSync("cat .git/ORIG_HEAD").toString();
    } catch (ignore) {
        console.warn("Failed to get commit sha by reading from ORIG_HEAD");
    }
    return null;
}

/*
*   Builders
*/

proGulp.task("buildMainHtml", () => {
    return gulp.src(`${APP_DIR}/main.html`)
        .pipe(gp.preprocess({context: {EXEC_ENV, NODE_ENV}}))
        .pipe(gp.rename("index.html"))
        .pipe(gulp.dest(`${BUILD_DIR}/`));
});

proGulp.task("buildAllScripts", (() => {
    var compiler = null;
    return trashCompiler => {
        if (trashCompiler) {
            compiler = null;
        }
        const deps = JSON.parse(fs.readFileSync(DEPS_PATH));
        fs.mkdirpSync(`${BUILD_DIR}/_assets/js`);
        compiler = compiler || webpack({
            entry: {
                app: `${APP_DIR}/main.jsx`,
                vendor: deps.js
            },
            devtool: "source-map",
            output: {
                filename: `${BUILD_DIR}/_assets/js/app.js`
            },
            module: {
                loaders: [
                    {
                        test: /\.jsx?$/,
                        exclude: /node_modules/,
                        loader: "babel"
                    },
                    {
                        test: /\.json$/,
                        loader: "json"
                    }
                ]
            },
            resolve: {
                root: APP_DIR,
                extensions: ["", ".js", ".json", ".jsx"]
            },
            plugins: [
                new webpack.DefinePlugin({
                    "process.env.NODE_ENV": JSON.stringify(NODE_ENV),
                    "process.env.EXEC_ENV": JSON.stringify(EXEC_ENV)
                }),
                new webpack.optimize.DedupePlugin(),
                new webpack.optimize.CommonsChunkPlugin(
                    "vendor",
                    `${BUILD_DIR}/_assets/js/vendor.js`
                ),
                (MINIFY_FILES ? new webpack.optimize.UglifyJsPlugin() : null)
            ].filter(i => i)
        });
        return promisify(compiler.run, {context: compiler})();
    };
})());

proGulp.task("buildAppAssets", () => {
    return gulp.src(`${APP_DIR}/assets/**/*`)
        .pipe(gulp.dest(`${BUILD_DIR}/_assets/`));
});

proGulp.task("buildAppVersion", () => {
    const pkg = JSON.parse(fs.readFileSync(`${process.cwd()}/package.json`, "utf8"));
    const commitSha = getCommitSha();
    const commitShaString = commitSha ? ` - ${commitSha}` : "";
    const version = `${pkg.version}${commitShaString}`;
    fs.writeFileSync(`${BUILD_DIR}/VERSION.txt`, version);
});

proGulp.task("buildAppChangelog", () => {
    return gulp.src(`${process.cwd()}/CHANGELOG.md`)
        .pipe(gulp.dest(`${BUILD_DIR}/`));
});

proGulp.task("buildVendorStyles", () => {
    const deps = JSON.parse(fs.readFileSync(DEPS_PATH));
    return gulp.src(deps.css)
        .pipe(gp.concat("vendor.css"))
        .pipe(gp.if(MINIFY_FILES, gp.cssnano()))
        .pipe(gulp.dest(`${BUILD_DIR}/_assets/css/`));
});

proGulp.task("buildVendorFonts", () => {
    const deps = JSON.parse(fs.readFileSync(DEPS_PATH));
    return gulp.src(deps.fonts)
        .pipe(gulp.dest(`${BUILD_DIR}/_assets/fonts/`));
});

/*
*   Exports
*/

const build = proGulp.parallel([
    "buildMainHtml",
    "buildAllScripts",
    "buildAppAssets",
    "buildAppVersion",
    "buildAppChangelog",
    "buildVendorStyles",
    "buildVendorFonts"
]);
Object.assign(build, {
    mainHtml: proGulp.task("buildMainHtml"),
    allScripts: proGulp.task("buildAllScripts"),
    appAssets: proGulp.task("buildAppAssets"),
    appVersion: proGulp.task("buildAppVersion"),
    appChangelog: proGulp.task("buildAppChangelog"),
    vendorStyles: proGulp.task("buildVendorStyles"),
    vendorFonts: proGulp.task("buildVendorFonts")
});
module.exports = build;


src/tasks/config.js
const dotenv = require("dotenv");
const fs = require("fs-extra");
const _ = require("lodash");
const proGulp = require("pro-gulp");

const {NODE_ENV, BUILD_DIR} = require("../config");

module.exports = proGulp.task("config", () => {
    var config = {};
    if (NODE_ENV === "development") {
        // In development, read from the `.env` file
        try {
            const env = fs.readFileSync(`${process.cwd()}/.env`);
            config = dotenv.parse(env);
        } catch (ignore) {
            console.log("Failed to read configuration from file `.env`");
        }
    }
    if (NODE_ENV === "production") {
        // In production, read from the `process.env` but only those variables
        // having a key that starts with __APP_CONFIG__
        const prefixRegexp = /^__APP_CONFIG__/;
        config = _(process.env)
            .pickBy((value, key) => prefixRegexp.test(key))
            .mapKeys((value, key) => key.replace(prefixRegexp, ""));
    }
    const code = `window.APP_CONFIG = ${JSON.stringify(config, null, 4)};`;
    fs.ensureDirSync(BUILD_DIR);
    fs.writeFileSync(`${BUILD_DIR}/app-config.js`, code);
});


gulpfile.babel.js
import {promisify} from "bluebird";
import browserSync from "browser-sync";
import {execSync, spawn} from "child_process";
import history from "connect-history-api-fallback";
import dotenv from "dotenv";
import fs from "fs";
import gulp from "gulp";
import gulpLoadPlugins from "gulp-load-plugins";
import _ from "lodash";
import mkdirp from "mkdirp";
import proGulp from "pro-gulp";
import webpack from "webpack";

const gp = gulpLoadPlugins();



/*
*   Constants
*/
const {NODE_ENV = "development", EXEC_ENV = "browser"} = process.env;
const MINIFY_FILES = (NODE_ENV === "production");
const testDir = `${process.cwd()}/test`;
const appDir = `${process.cwd()}/app`;
const buildDir = `${process.cwd()}/build`;
const depsPath = `${process.cwd()}/deps.json`;
const packagePath = `${process.cwd()}/package.json`;
const npmDir = `${process.cwd()}/node_modules/.bin`;



/*
*   Utils
*/

function getCommitSha () {
    try {
        return execSync("git rev-parse HEAD").toString();
    } catch (ignore) {
        console.warn("Failed to get commit sha via git command");
    }
    try {
        return execSync("cat .git/ORIG_HEAD").toString();
    } catch (ignore) {
        console.warn("Failed to get commit sha by reading from ORIG_HEAD");
    }
    return null;
}



/*
*   Builders
*/

proGulp.task("buildMainHtml", () => {
    return gulp.src(`${appDir}/main.html`)
        .pipe(gp.preprocess({context: {EXEC_ENV, NODE_ENV}}))
        .pipe(gp.rename("index.html"))
        .pipe(gulp.dest(`${buildDir}/`));
});

proGulp.task("buildAllScripts", (() => {
    const deps = JSON.parse(fs.readFileSync(depsPath));
    mkdirp.sync(`${buildDir}/_assets/js`);
    const compiler = webpack({
        entry: {
            app: `${appDir}/main.jsx`,
            vendor: deps.js
        },
        devtool: "source-map",
        output: {
            filename: `${buildDir}/_assets/js/app.js`
        },
        module: {
            loaders: [
                {
                    test: /\.jsx?$/,
                    exclude: /node_modules/,
                    loader: "babel"
                },
                {
                    test: /\.json$/,
                    loader: "json"
                }
            ]
        },
        resolve: {
            root: appDir,
            extensions: ["", ".js", ".json", ".jsx"]
        },
        plugins: [
            new webpack.DefinePlugin({
                "process.env.NODE_ENV": JSON.stringify(NODE_ENV),
                "process.env.EXEC_ENV": JSON.stringify(EXEC_ENV)
            }),
            new webpack.optimize.DedupePlugin(),
            new webpack.optimize.CommonsChunkPlugin(
                "vendor",
                `${buildDir}/_assets/js/vendor.js`
            ),
            (MINIFY_FILES ? new webpack.optimize.UglifyJsPlugin() : null)
        ].filter(i => i)
    });
    return promisify(::compiler.run);
})());

proGulp.task("buildAppAssets", () => {
    return gulp.src(`${appDir}/assets/**/*`)
        .pipe(gulp.dest(`${buildDir}/_assets/`));
});

proGulp.task("buildAppVersion", () => {
    const pkg = JSON.parse(fs.readFileSync(`${process.cwd()}/package.json`, "utf8"));
    const commitSha = getCommitSha();
    const commitShaString = commitSha ? ` - ${commitSha}` : "";
    const version = `${pkg.version}${commitShaString}`;
    fs.writeFileSync(`${buildDir}/VERSION`, version);
});

proGulp.task("buildVendorStyles", () => {
    const deps = JSON.parse(fs.readFileSync(depsPath));
    return gulp.src(deps.css)
        .pipe(gp.concat("vendor.css"))
        .pipe(gp.if(MINIFY_FILES, gp.cssnano()))
        .pipe(gulp.dest(`${buildDir}/_assets/css/`));
});

proGulp.task("buildVendorFonts", () => {
    const deps = JSON.parse(fs.readFileSync(depsPath));
    return gulp.src(deps.fonts)
        .pipe(gulp.dest(`${buildDir}/_assets/fonts/`));
});

proGulp.task("build", proGulp.parallel([
    "buildMainHtml",
    "buildAllScripts",
    "buildAppAssets",
    "buildAppVersion",
    "buildVendorStyles",
    "buildVendorFonts"
]));

gulp.task("build", proGulp.task("build"));



/*
*   Config generator
*/

proGulp.task("config", () => {
    var config = {};
    if (NODE_ENV === "development") {
        // In development, read from the `.env` file
        try {
            const env = fs.readFileSync(`${process.cwd()}/.env`);
            config = dotenv.parse(env);
        } catch (ignore) {
            console.log("Failed to read configuration from file `.env`");
        }
    }
    if (NODE_ENV === "production") {
        // In production, read from the `process.env` but only those variables
        // having a key that starts with __APP_CONFIG__
        const prefixRegexp = /^__APP_CONFIG__/;
        config = _(process.env)
            .pickBy((value, key) => prefixRegexp.test(key))
            .mapKeys((value, key) => key.replace(prefixRegexp, ""));
    }
    const code = `window.APP_CONFIG = ${JSON.stringify(config, null, 4)};`;
    fs.writeFileSync(`${buildDir}/app-config.js`, code);
});

gulp.task("config", proGulp.task("config"));



/*
*   Linter
*/

gulp.task("lint", () => {
    const srcs = [
        `${appDir}/**/*.js`,
        `${appDir}/**/*.jsx`,
        `!${appDir}/assets/**/*`,
        `${testDir}/**/*.js`,
        `${testDir}/**/*.jsx`
    ];
    return gulp.src(srcs)
        .pipe(gp.eslint())
        .pipe(gp.eslint.format())
        .pipe(gp.eslint.failAfterError());
});



/*
 *   Retire
 */

gulp.task("retire", function () {
    // Spawn Retire.js as a child process
    // You can optionally add option parameters to the second argument (array)
    var child = spawn("retire", [], {cwd: process.cwd()});

    child.stdout.setEncoding("utf8");
    child.stdout.on("data", function (data) {
        gp.util.log(data);
    });

    child.stderr.setEncoding("utf8");
    child.stderr.on("data", function (data) {
        gp.util.log(gp.util.colors.red(data));
        gp.util.beep();
    });
});

gulp.task("retire", proGulp.task("retire"));

/*
*   Testers
*/

proGulp.task("test", () => {
    return gulp.src([`${testDir}/**/*.js`, `${testDir}/**/*.jsx`])
        .pipe(gp.spawnMocha({
            compilers: "jsx:babel-register",
            env: {
                NODE_ENV: "test",
                NODE_PATH: `${appDir}`
            }
        }))
        .on("error", function () {
            // Swallow errors
            this.emit("end");
        });
});

gulp.task("test", proGulp.task("test"));

proGulp.task("coverage", () => {
    const command = [
        "env NODE_ENV=\"test\"",
        `env NODE_PATH="${appDir}"`,
        `${npmDir}/babel-node`,
        `${npmDir}/isparta cover --include "**/*.js" --include "**/*.jsx"`,
        `${npmDir}/_mocha -- test/*.js* test/**/*.js*`
    ].join(" ");
    execSync(command, {
        env: process.env,
        stdio: "inherit"
    });
});

gulp.task("coverage", proGulp.task("coverage"));



/*
*   Tasks to setup the development environment
*/

proGulp.task("setupDevServer", () => {
    browserSync({
        server: {
            baseDir: buildDir,
            middleware: [history()]
        },
        files: `${buildDir}/**/*`,
        port: 8080,
        ghostMode: false,
        injectChanges: false,
        notify: false,
        open: false,
        reloadDebounce: 1000
    });
});

proGulp.task("setupWatchers", () => {
    gulp.watch(
        `${appDir}/main.html`,
        proGulp.task("buildMainHtml")
    );
    gulp.watch(
        `${appDir}/.env`,
        proGulp.task("config")
    );
    gulp.watch(
        [`${appDir}/**/*.jsx`, `${appDir}/**/*.js`],
        proGulp.parallel(["buildAllScripts", "test"])
    );
    gulp.watch(
        `${appDir}/assets/**/*`,
        proGulp.task("buildAppAssets")
    );
    gulp.watch(
        [`${testDir}/**/*.jsx`, `${testDir}/**/*.js`],
        proGulp.task("test")
    );
    gulp.watch(
        depsPath,
        proGulp.parallel(["buildAllScripts", "buildVendorFonts", "buildVendorStyles", "test"])
    );
    gulp.watch(
        [`${appDir}/**/*.js`, packagePath],
        proGulp.task("retire")
    );
});

gulp.task("dev", proGulp.sequence([
    "build",
    "config",
    "test",
    "retire",
    "setupDevServer",
    "setupWatchers"
]));



/*
*   Default task
*/

gulp.task("default", () => {
    gp.util.log("");
    gp.util.log("Usage: " + gp.util.colors.blue("sd-builder [TASK]"));
    gp.util.log("");
    gp.util.log("Available tasks:");
    gp.util.log("  " + gp.util.colors.green("build") + "    build the project");
    gp.util.log("  " + gp.util.colors.green("config") + "   write the configuration to app-config.js");
    gp.util.log("  " + gp.util.colors.green("dev") + "      set up dev environment with auto-recompiling");
    gp.util.log("  " + gp.util.colors.green("lint") + "     lint application source code");
    gp.util.log("  " + gp.util.colors.green("test") + "     run tests");
    gp.util.log("  " + gp.util.colors.green("coverage") + " run tests and calculate coverage");
    gp.util.log("  " + gp.util.colors.green("retire") + "   check dependencies vulnerabilities");
    gp.util.log("");
});
